core/src/main/java/org/keycloak/Feature.java

```diff
@@ -0,0 +1,245 @@
+     *     - distribution/feature-packs/server-feature-pack/src/main/resources/content/bin/migrate-domain-clustered.cli
+     *     - distribution/distribution-tests/src/test/java/org/keycloak/test/distribution/DistributionFeaturesTest.java  (COMMUNITY_FEATURES, and PRODUCT_FEATURES constants)
+     *
+     *   In addition, the following files may need to be modified to support tests:
+     *     - testsuite/integration-arquillian/tests/base/src/test/resources/META-INF/keycloak-server.json (feature section)
+     *     - testsuite/utils/src/main/resources/META-INF/keycloak-server.json (feature section)
+     *
+     */
+
+    AUTHORIZATION(true), IMPERSONATION, SCRIPTS(true), DOCKER, ACCOUNT2(true), TOKEN_EXCHANGE;
+
+
+    private static final Logger log = Logger.getLogger(Feature.class.getName());
+
+    private boolean preview;
+
+    private static boolean foundBadConfig = false;
+
+    Feature() {
+        this.preview = false;
+    }
+
+    Feature(boolean preview) {
+        this.preview = preview;
+    }
+
+    public String caption() {
+        String s = name().toLowerCase().replace('_', '-');
+        return preview ? s + "-preview" : s;
+    }
+
+    public static Set<Feature> getDisabledFeatures() {
+        // by default all features are disabled
+        HashSet<Feature> disabled = new HashSet(Arrays.asList(values()));
+        Config.Scope config = Config.scope("feature");
+        for (Feature f: values()) {
+            Config.Scope featureScope = config.scope(f.caption());
+            if (!featureScope.getBoolean("enabled", false)) {
+                disabled.add(f);
+            } else {
+                disabled.remove(f);
+            }
+        }
+
+        return disabled;
+    }
+
+    public static boolean isFeatureEnabled(Feature feature) {
+        return Config.scope("feature", feature.caption()).getBoolean("enabled", false);
+    }
+
+    public static Feature fromCaption(String name) {
+        Feature ret = AUTHORIZATION.caption().equals(name) ? AUTHORIZATION :
+                IMPERSONATION.caption().equals(name) ? IMPERSONATION :
+                SCRIPTS.caption().equals(name) ? SCRIPTS :
+                DOCKER.caption().equals(name) ? DOCKER :
+                ACCOUNT2.caption().equals(name) ? ACCOUNT2 :
+                TOKEN_EXCHANGE.caption().equals(name) ? TOKEN_EXCHANGE : null;
+
+        if (ret == null) {
+            throw new IllegalArgumentException("No such feature: " + name);
+        }
+        return ret;
+    }
+
+    public static List<String> validCaptions() {
+        Feature[] features = values();
+        List<String> ret = new ArrayList<>(features.length);
+
+        for (Feature f: features) {
+            ret.add(f.caption());
+        }
+        return ret;
+    }
+
+    public boolean isEnabledByDefault() {
+        // Value is dependent on whether this is a Keycloak release or some other
```

Comment:
Would be better to add a constructor for enums to drive this. Something like:

Feature(boolean enabledByDefault, boolen preview)

Basically, it would be good to only have to update the values for the feature and not have to edit methods themselves.

Reply:
I overlooked the fact that this part is independent of Config so the only factor here is Version.NAME which is a static value. Will change like you suggest.

Reply:
BTW: I noticed that Version.XXX constants are almost all non-final. The reason is how they are initialised. Can I fix that as part of this issue - I want to add IS_COMMUNITY_VERSION constant to Version class? Or I can make a separate PR for that.

Reply:
You can change the Version in this PR if you want